### ISSUE TITLE
How to start training unlabelled data from scratch

### ISSUE THREAD

**Issue by harveyng** [wvangansbeke/Unsupervised-Classification#15]
Hi, thanks for an amazing paper and code.

I have a question regarding how to start a brand new dataset with all unlabeled-images. 

I follow your suggestion on other issues, change configs/env.yml, 
create configs/pretext/simclr_*.yml, 
update, utils/common_config
create new class for custom dataset,

I wonder how i should deal with the validation step as i dont have any label in place, could i just edit out the _val_db_name_: in the pretext/yml file or do i need to do other changes. Any advise is greatly appreciated.

**Comment by wvangansbeke** [comment]
Hi @harveyng,
If you can not validate for your own dataset, then you can indeed comment out the validation part or remove it. It should be fairly obvious from the code.

**Comment by wvangansbeke** [comment]
If something is still unclear, let me know. I will close this for now.

**Comment by harveyng** [comment]
Thanks @wvangansbeke , i finished train first step using simclr. However i still have trouble running second step.


   ```
 for epoch in range(start_epoch, p['epochs']):
        print(colored('Epoch %d/%d' %(epoch+1, p['epochs']), 'yellow'))
        print(colored('-'*15, 'yellow'))

        lr = adjust_learning_rate(p, optimizer, epoch)
        print('Adjusted learning rate to {:.5f}'.format(lr))

        # Train
        print('Train ...')
        scan_train(train_dataloader, model, criterion, optimizer, epoch, p['update_cluster_head_only'])

        # Evaluate 
        print('Make prediction on validation set ...')
        predictions = get_predictions(p, val_dataloader, model)

        print('Evaluate based on SCAN loss ...')
        scan_stats = scan_evaluate(predictions)
        print(scan_stats)
        lowest_loss_head = scan_stats['lowest_loss_head']
        lowest_loss = scan_stats['lowest_loss']
       
        if lowest_loss < best_loss:
            print('New lowest loss on validation set: %.4f -> %.4f' %(best_loss, lowest_loss))
            print('Lowest loss head is %d' %(lowest_loss_head))
            best_loss = lowest_loss
            best_loss_head = lowest_loss_head
            torch.save({'model': model.module.state_dict(), 'head': best_loss_head}, p['scan_model'])

        else:
            print('No new lowest loss on validation set: %.4f -> %.4f' %(best_loss, lowest_loss))
            print('Lowest loss head is %d' %(best_loss_head))

        print('Evaluate with hungarian matching algorithm ...')
        clustering_stats = hungarian_evaluate(lowest_loss_head, predictions, compute_confusion_matrix=False)
        print(clustering_stats)     

        # Checkpoint
        print('Checkpoint ...')
        torch.save({'optimizer': optimizer.state_dict(), 'model': model.state_dict(), 
                    'epoch': epoch + 1, 'best_loss': best_loss, 'best_loss_head': best_loss_head},
                     p['scan_checkpoint'])
    
    # Evaluate and save the final model
    print(colored('Evaluate best model based on SCAN metric at the end', 'blue'))
    model_checkpoint = torch.load(p['scan_model'], map_location='cpu')
    model.module.load_state_dict(model_checkpoint['model'])
    predictions = get_predictions(p, val_dataloader, model)
    clustering_stats = hungarian_evaluate(model_checkpoint['head'], predictions, 
                            class_names=val_dataset.dataset.classes, 
                            compute_confusion_matrix=True, 
                            confusion_matrix_file=os.path.join(p['scan_dir'], 'confusion_matrix.png'))
    print(clustering_stats)  
```

This required the val_dataloader for evaluating the result and save the model base on that. Can you advise how should i change the code when i dont have any validation dataset? Same with the final step selflabel as well.

Thanks

**Comment by wvangansbeke** [comment]
Hi @harveyng,

You can probably just comment it out, but it is hard to guarantee that everything will work smoothly for your particular problem. Maybe it's better to first visualize the nearest neighbors after training Simclr  and check their quality. 

Wouter

**Comment by ggndtes** [comment]
Hi @harveyng, would you be able the progress you have been making with running it using your own dataset?

Thx

**Comment by harveyng** [comment]
> Hi @harveyng,
> 
> You can probably just comment it out, but it is hard to guarantee that everything will work smoothly for your particular problem. Maybe it's better to first visualize the nearest neighbors after training Simclr and check their quality.
> 
> Wouter

Thanks @wvangansbeke , i thought so, great approach overall. I probably run SIMCLR and use it to label some data then turn into a semi-supervised problem.

**Comment by harveyng** [comment]
> Hi @harveyng, would you be able the progress you have been making with running it using your own dataset?
> 
> Thx

Hi @ggndtes, I still experiment with my dataset. But as wvangansbeke mentioned, you can try it step by step and check the quality a long the way.

**Comment by DalalAlmutawa** [comment]
> > Hi @harveyng, would you be able the progress you have been making with running it using your own dataset?
> > Thx
> 
> Hi @ggndtes, I still experiment with my dataset. But as wvangansbeke mentioned, you can try it step by step and check the quality a long the way.

Please mention
How visualize the nearest neighbors after training?

**Comment by DalalAlmutawa** [comment]
Please share with me the DataLoad code for your data set
